### PR TITLE
Remove unused function codes

### PR DESCRIPTION
**Pull request by richsalz**
Output of util/find-unused-funcs

### CODE REVIEW

**Comment by richsalz** [comment]
an updated commit avoids missing with ERR_FATAL_ERROR

File: crypto/rsa/rsa_err.c [hunk 1]
```diff
@@ -162,14 +161,11 @@ static ERR_STRING_DATA RSA_str_reasons[] = {
...
     {ERR_REASON(RSA_R_UNKNOWN_DIGEST), "unknown digest"},
-    {ERR_REASON(RSA_R_UNKNOWN_MASK_DIGEST), "unknown mask digest"},
     {ERR_REASON(RSA_R_UNKNOWN_PADDING_TYPE), "unknown padding type"},
     {ERR_REASON(RSA_R_UNSUPPORTED_ENCRYPTION_TYPE),
      "unsupported encryption type"},
     {ERR_REASON(RSA_R_UNSUPPORTED_LABEL_SOURCE), "unsupported label source"},
-    {ERR_REASON(RSA_R_UNSUPPORTED_MASK_ALGORITHM),
      "unsupported mask algorithm"},
-    {ERR_REASON(RSA_R_UNSUPPORTED_MASK_PARAMETER),
      "unsupported mask parameter"},
```
Comment (levitte) [inline crypto/rsa/rsa_err.c]: Uhmmm, didn't notice this earlier.  Do you think you could remove the second line too?  ;-)

File: crypto/rsa/rsa_err.c [hunk 2]
```diff
@@ -162,14 +161,11 @@ static ERR_STRING_DATA RSA_str_reasons[] = {
      "the asn1 object identifier is not known for this md"},
     {ERR_REASON(RSA_R_UNKNOWN_ALGORITHM_TYPE), "unknown algorithm type"},
     {ERR_REASON(RSA_R_UNKNOWN_DIGEST), "unknown digest"},
-    {ERR_REASON(RSA_R_UNKNOWN_MASK_DIGEST), "unknown mask digest"},
     {ERR_REASON(RSA_R_UNKNOWN_PADDING_TYPE), "unknown padding type"},
     {ERR_REASON(RSA_R_UNSUPPORTED_ENCRYPTION_TYPE),
      "unsupported encryption type"},
     {ERR_REASON(RSA_R_UNSUPPORTED_LABEL_SOURCE), "unsupported label source"},
-    {ERR_REASON(RSA_R_UNSUPPORTED_MASK_ALGORITHM),
      "unsupported mask algorithm"},
```
Comment (levitte) [inline crypto/rsa/rsa_err.c]: Here too.

File: crypto/x509/x509_err.c [hunk 3]
```diff
@@ -72,7 +72,6 @@ static ERR_STRING_DATA X509_str_functs[] = {
     {ERR_FUNC(X509_F_X509_REQ_TO_X509), "X509_REQ_to_X509"},
     {ERR_FUNC(X509_F_X509_STORE_ADD_CERT), "X509_STORE_add_cert"},
     {ERR_FUNC(X509_F_X509_STORE_ADD_CRL), "X509_STORE_add_crl"},
-    {ERR_FUNC(X509_F_X509_STORE_CTX_GET1_ISSUER),
      "X509_STORE_CTX_get1_issuer"},
```
Comment (levitte) [inline crypto/x509/x509_err.c]: ... and here

File: ssl/ssl_err.c [hunk 4]
```diff
@@ -59,19 +57,15 @@ static ERR_STRING_DATA SSL_str_functs[] = {
     {ERR_FUNC(SSL_F_FINAL_SIG_ALGS), "final_sig_algs"},
     {ERR_FUNC(SSL_F_NSS_KEYLOG_INT), "nss_keylog_int"},
     {ERR_FUNC(SSL_F_OPENSSL_INIT_SSL), "OPENSSL_init_ssl"},
-    {ERR_FUNC(SSL_F_OSSL_STATEM_CLIENT13_READ_TRANSITION),
      "ossl_statem_client13_read_transition"},
```
Comment (levitte) [inline ssl/ssl_err.c]: ... and ...  you know, I'll stop here and trust you'll find the rest of them

**Comment by richsalz** [comment]
There are a bunch of these I didn’t clean up ☹  fixup commit coming

**Comment by richsalz** [comment]
updated commit pushed.  and it builds.

**Comment by mattcaswell** [comment]
Err...isn't this an API compat issue? This is removing defines from public headers

**Comment by levitte** [comment]
Unfortunately, @mattcaswell has a point...

**Comment by richsalz** [comment]
Almost all of these changes are updating function names, the ```_F_```  defines.  There are 50 of those.  They are not part of our documented API.

There are 18 unused reason codes:
```
; g l1 | grep -- '-#.*_R_'
-# define EVP_R_ERROR_SETTING_FIPS_MODE                    166
-# define RSA_R_UNKNOWN_MASK_DIGEST                        151
-# define RSA_R_UNSUPPORTED_MASK_ALGORITHM                 153
-# define RSA_R_UNSUPPORTED_MASK_PARAMETER                 154
-# define SSL_R_AT_LEAST_TLS_1_0_NEEDED_IN_FIPS_MODE       143
-# define SSL_R_BAD_HELLO_REQUEST                          105
-# define SSL_R_BAD_PACKET_LENGTH                          115
-# define SSL_R_EMPTY_SRTP_PROTECTION_PROFILE_LIST         354
-# define SSL_R_NO_SHARED_GROUPS                           410
-# define SSL_R_RENEGOTIATE_EXT_TOO_LONG                   335
-# define SSL_R_SERVERHELLO_TLSEXT                         275
-# define SSL_R_SRTP_PROTECTION_PROFILE_LIST_TOO_LONG      363
-# define SSL_R_SSL_NEGATIVE_LENGTH                        372
-# define SSL_R_TLS_HEARTBEAT_PEER_DOESNT_ACCEPT           365
-# define SSL_R_TLS_HEARTBEAT_PENDING                      366
-# define SSL_R_UNKNOWN_PKEY_TYPE                          251
-# define SSL_R_USE_SRTP_NOT_NEGOTIATED                    369
-# define X509_R_SHOULD_RETRY                              106
```

I can turn those into "define ... 0" if you prefer. Or we can remove them and see if anyone complains.

**Comment by mattcaswell** [comment]
I don't think we can remove them (unless there are any that are unused which only exist in master). We should not set them to 0 either. That will mean that when we add new reason codes we will reuse the old numbers. That in theory could cause problems for binary compat (if someone was using one of the numbers to match a specific error then suddenly they will be matching a different error if they change the library version without recompiling their app).

**Comment by richsalz** [comment]
I don't think that's right about the error codes, it assigns the next higher one, but fine.  ```SSL_R_NO_SHARED_GROUPS``` is the only reason code that was not in 1.1.0  So I will restore all the others and push a new commit.

**Comment by richsalz** [comment]
new commits pushed to address the api compatibility issues.  ping @levitte and @mattcaswell

**Comment by richsalz** [comment]
All error codes restored, just has internal function codes now.  ping @mattcaswell @levitte

**Comment by richsalz** [comment]
ping

**Comment by pauli-oracle** [comment]
Would changing these codes from #defines to extern const int's help with future compatibility issues?  Number reuse would cease to be a problem.  In fact the files could be generated by the build scripts and not live in git which would help with merge conflicts.  The big downsides are no more #ifdef checks or switch cases.

Anyway, just a thought.

**Comment by richsalz** [comment]
So given @mattcaswell 's comments on https://github.com/openssl/openssl/pull/2244 I guess this is a non-starter?  Really?  Ping @openssl/dev

**Comment by kaduk** [comment]
Applications may well have code that inspects the error stack and uses these symbols in conditionals.  Those conditionals are now stale, but if we remove the symbols, those applications fail to compile.
I might be able to convince myself that it's okay to define all the function codes for removed functions to the same invalid value, as I don't think we claim to support compiling against a newer codebase than is used at runtime.

**Comment by levitte** [comment]
I would say that this is ok *as long as it only removes codes that don't exist in 1.1.0*

**Comment by mattcaswell** [comment]
> I would say that this is ok as long as it only removes codes that don't exist in 1.1.0

Right. If they were added in 1.1.1 and are now no longer needed then we're fine to remove them again.

File: include/openssl/bio.h [hunk 5]
```diff
@@ -810,17 +810,14 @@ int ERR_load_BIO_strings(void);
...
 # define BIO_F_BIO_READ_INTERN                            120
 # define BIO_F_BIO_SOCKET                                 140
 # define BIO_F_BIO_SOCKET_NBIO                            142
 # define BIO_F_BIO_SOCK_INFO                              141
 # define BIO_F_BIO_SOCK_INIT                              112
 # define BIO_F_BIO_WRITE                                  113
-# define BIO_F_BIO_WRITE_EX                               119
 # define BIO_F_BIO_WRITE_INTERN                           128
 # define BIO_F_BUFFER_CTRL                                114
-# define BIO_F_CONN_CTRL                                  127
```
Comment (levitte) [inline include/openssl/bio.h]: Undo this line

File: include/openssl/rsa.h [hunk 6]
```diff
@@ -500,7 +500,6 @@ int ERR_load_RSA_strings(void);
 # define RSA_F_RSA_METH_DUP                               161
 # define RSA_F_RSA_METH_NEW                               162
 # define RSA_F_RSA_METH_SET1_NAME                         163
-# define RSA_F_RSA_MGF1_TO_MD                             157
```
Comment (levitte) [inline include/openssl/rsa.h]: Undo this line

File: include/openssl/ssl.h [hunk 7]
```diff
@@ -2147,7 +2146,6 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_DO_SSL3_WRITE                              104
 # define SSL_F_DTLS1_BUFFER_RECORD                        247
 # define SSL_F_DTLS1_CHECK_TIMEOUT_NUM                    318
-# define SSL_F_DTLS1_HEARTBEAT                            305
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 8]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_SSL3_WRITE_BYTES                           158
 # define SSL_F_SSL3_WRITE_PENDING                         159
 # define SSL_F_SSL_ADD_CERT_CHAIN                         316
-# define SSL_F_SSL_ADD_CERT_TO_BUF                        319
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 9]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_ADD_CERT_TO_WPACKET                    493
-# define SSL_F_SSL_ADD_CLIENTHELLO_RENEGOTIATE_EXT        298
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 10]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_ADD_CLIENTHELLO_TLSEXT                 277
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 11]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_ADD_CLIENTHELLO_USE_SRTP_EXT           307
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 12]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_ADD_DIR_CERT_SUBJECTS_TO_STACK         215
 # define SSL_F_SSL_ADD_FILE_CERT_SUBJECTS_TO_STACK        216
-# define SSL_F_SSL_ADD_SERVERHELLO_RENEGOTIATE_EXT        299
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 13]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_ADD_SERVERHELLO_TLSEXT                 278
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 14]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_ADD_SERVERHELLO_USE_SRTP_EXT           308
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 15]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_BAD_METHOD                             160
 # define SSL_F_SSL_BUILD_CERT_CHAIN                       332
-# define SSL_F_SSL_BYTES_TO_CIPHER_LIST                   161
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 16]
```diff
@@ -2198,26 +2191,17 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_CACHE_CIPHERLIST                       520
 # define SSL_F_SSL_CERT_ADD0_CHAIN_CERT                   346
 # define SSL_F_SSL_CERT_DUP                               221
 # define SSL_F_SSL_CERT_NEW                               162
 # define SSL_F_SSL_CERT_SET0_CHAIN                        340
 # define SSL_F_SSL_CHECK_PRIVATE_KEY                      163
-# define SSL_F_SSL_CHECK_SERVERHELLO_TLSEXT               280
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 17]
```diff
@@ -2226,7 +2210,6 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_SSL_COMP_ADD_COMPRESSION_METHOD            165
 # define SSL_F_SSL_CONF_CMD                               334
 # define SSL_F_SSL_CREATE_CIPHER_LIST                     166
-# define SSL_F_SSL_CTRL                                   232
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 18]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_SSL_ENABLE_CT                              402
 # define SSL_F_SSL_GET_NEW_SESSION                        181
 # define SSL_F_SSL_GET_PREV_SESSION                       217
-# define SSL_F_SSL_GET_SERVER_CERT_INDEX                  322
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 19]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_GET_SIGN_PKEY                          183
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 20]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_INIT_WBIO_BUFFER                       184
 # define SSL_F_SSL_KEY_UPDATE                             515
 # define SSL_F_SSL_LOAD_CLIENT_CA_FILE                    185
-# define SSL_F_SSL_LOG_MASTER_SECRET                      498
 # define SSL_F_SSL_LOG_RSA_CLIENT_KEY_EXCHANGE            499
 # define SSL_F_SSL_MODULE_INIT                            392
 # define SSL_F_SSL_NEW                                    186
-# define SSL_F_SSL_PARSE_CLIENTHELLO_RENEGOTIATE_EXT      300
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 21]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_CLIENTHELLO_TLSEXT               302
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 22]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_CLIENTHELLO_USE_SRTP_EXT         310
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 23]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_SERVERHELLO_RENEGOTIATE_EXT      301
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 24]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_SERVERHELLO_TLSEXT               303
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 25]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_SERVERHELLO_USE_SRTP_EXT         311
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 26]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
 # define SSL_F_SSL_PEEK                                   270
-# define SSL_F_SSL_PEEK_EX                                432
 # define SSL_F_SSL_PEEK_INTERNAL                          522
 # define SSL_F_SSL_READ                                   223
 # define SSL_F_SSL_READ_EARLY_DATA                        529
-# define SSL_F_SSL_READ_EX                                434
 # define SSL_F_SSL_READ_INTERNAL                          523
 # define SSL_F_SSL_RENEGOTIATE                            516
-# define SSL_F_SSL_SCAN_CLIENTHELLO_TLSEXT                320
-# define SSL_F_SSL_SCAN_SERVERHELLO_TLSEXT                321
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 27]
```diff
@@ -2257,31 +2240,18 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_PARSE_SERVERHELLO_USE_SRTP_EXT         311
 # define SSL_F_SSL_PEEK                                   270
-# define SSL_F_SSL_PEEK_EX                                432
 # define SSL_F_SSL_PEEK_INTERNAL                          522
 # define SSL_F_SSL_READ                                   223
 # define SSL_F_SSL_READ_EARLY_DATA                        529
-# define SSL_F_SSL_READ_EX                                434
 # define SSL_F_SSL_READ_INTERNAL                          523
 # define SSL_F_SSL_RENEGOTIATE                            516
-# define SSL_F_SSL_SCAN_CLIENTHELLO_TLSEXT                320
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

**Comment by richsalz** [comment]
doing that changes from 52 functions to 26.  removing 26 string table entries as well.
is that worth doing?

File: include/openssl/crypto.h [hunk 28]
```diff
@@ -442,7 +442,6 @@ int ERR_load_CRYPTO_strings(void);
 
 /* Function codes. */
 # define CRYPTO_F_CRYPTO_DUP_EX_DATA                      110
-# define CRYPTO_F_CRYPTO_FREE_EX_DATA                     111
```
Comment (levitte) [inline include/openssl/crypto.h]: Undo this line

**Comment by kaduk** [comment]
Seems like useful cleanup to me.

File: include/openssl/ssl.h [hunk 29]
```diff
@@ -2317,16 +2287,13 @@ int ERR_load_SSL_strings(void);
...
-# define SSL_F_SSL_WRITE_EARLY_FINISH                     527
-# define SSL_F_SSL_WRITE_EX                               433
 # define SSL_F_SSL_WRITE_INTERNAL                         524
 # define SSL_F_STATE_MACHINE                              353
 # define SSL_F_TLS12_CHECK_PEER_SIGALG                    333
 # define SSL_F_TLS12_COPY_SIGALGS                         533
 # define SSL_F_TLS13_CHANGE_CIPHER_STATE                  440
 # define SSL_F_TLS13_SETUP_KEY_BLOCK                      441
 # define SSL_F_TLS1_CHANGE_CIPHER_STATE                   209
-# define SSL_F_TLS1_CHECK_DUPLICATE_EXTENSIONS            341
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 30]
```diff
@@ -2351,16 +2318,12 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_TLS_CONSTRUCT_CLIENT_CERTIFICATE           484
 # define SSL_F_TLS_CONSTRUCT_CLIENT_HELLO                 487
 # define SSL_F_TLS_CONSTRUCT_CLIENT_KEY_EXCHANGE          488
-# define SSL_F_TLS_CONSTRUCT_CLIENT_VERIFY                489
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

File: include/openssl/ssl.h [hunk 31]
```diff
@@ -2427,7 +2381,6 @@ int ERR_load_SSL_strings(void);
 # define SSL_F_TLS_POST_PROCESS_CLIENT_KEY_EXCHANGE       384
 # define SSL_F_TLS_PREPARE_CLIENT_CERTIFICATE             360
 # define SSL_F_TLS_PROCESS_CERTIFICATE_REQUEST            361
-# define SSL_F_TLS_PROCESS_CERT_STATUS                    362
```
Comment (levitte) [inline include/openssl/ssl.h]: Undo this line

**Comment by levitte** [comment]
It seems I did double work.  Ah well...

**Comment by levitte** [comment]
I think this is a worthwhile effort for 1.2.0, so keep the script around in a PR, mark it for 1.2.0

**Comment by richsalz** [comment]
The script is in util/find-unused-errors.  Closing this.  sigh.

**Pull request closed without merging**